### DISCUSSION TITLE
Data tables must be first in where block?

### DISCUSSION THREAD

**Discussion by dmurat** [spockframework/spock#1424] [Q&A] [answered]
I will like to combine data variable assignment, data pipes, and data tables in the test. And this works when the data table is the first one in the `where` block:

```
  void "data assignment and data table sample 1"() {
    expect:
    a + b == d

    where:
    a | b
    1 | a + 1
    1 | a + 2
    1 | a + 3

    c = a + b
    d = c
  }
```

However, sometimes I will like that data assignment is before the data table:

```
  void "data pipe and data table sample 2"() {
    expect:
    a + b == d

    where:
    a = 1

    b     | _
    a + 1 | _
    a + 2 | _
    a + 3 | _

    c = a + b
    d = c
  }
```

Is this possible without declaring variable `a` as `static` or `Shared`?
I'm using Spock 2 with Groovy 3.

Tnx

**Comment by leonard84** [comment, marked as answer]
It is as you've described. It is not possible to change the order and you'll have to use `static` or `@Shared`.

**Comment by dmurat** [comment]
Wow, I finally found something Spock can't do 😄 
If I may suggest, I think it would be useful to state this explicitly in the documentation. Reading the section https://spockframework.org/spock/docs/2.0/data_driven_testing.html#_combining_data_tables_data_pipes_and_variable_assignments suggests that any combination is allowed.

**Reply by leonard84** [reply]
https://spockframework.org/spock/docs/2.0/data_driven_testing.html#_accessing_other_data_variables is even more confusing, but `c` isn't used in the table below and it also wouldn't work.

**Reply by dmurat** [reply]
Yes, I saw that one too, but just wanted to solve one puzzle at a time 😃 
Thank you for your confirmation.
